linker/shared/src/main/scala/org/scalajs/linker/frontend/modulesplitter/InternalModuleIDGenerator.scala

```diff
@@ -62,22 +88,79 @@ private[modulesplitter] final class InternalModuleIDGenerator(
+     * uppercase (and titlecase) code points can case-fold to the same
+     * lowercase letter. Therefore, the complete scheme is:
+     *
+     * - ASCII uppercase letters are prefixed with '-'.
+     * - '-' is prefixed with '-'.
+     * - Non-ASCII characters are all prefixed by '-u' followed by the 6
+     *   hexdigits of their codepoint.
+     *
+     * The last rule is far from being optimal, but it is safe. Encountering
+     * non-ASCII characters in class names should be rare anyway.
      */
-    var moduleID = ModuleID(name.nameString)
-    while (avoidSet.contains(moduleID))
-      moduleID = ModuleID(moduleID.id + ".")
-    moduleID
+
+    val builder = new java.lang.StringBuilder
+
+    // First, encode uppercase characters to avoid accidental case-insensitive clashes
+    val originalNameString = name.nameString
+    val originalNameStringLen = originalNameString.length()
+    var i = 0
+    while (i != originalNameStringLen) {
+      val cp = originalNameString.codePointAt(i)
+      if (cp < 0x80) {
+        // ASCII
+        if (cp == '-' || (cp >= 'A' && cp <= 'Z'))
+          builder.append('-')
+        builder.append(cp.toChar)
+        i += 1
+      } else {
+        // Non-ASCII
+        new java.util.Formatter(builder).format("-u%06x", Integer.valueOf(cp))
+        builder.appendCodePoint(cp)
+        i += Character.charCount(cp)
+      }
+    }
+
+    // Second, avoid colliding with the public module IDs in `avoidSet`
+    var candidateID = builder.toString()
+    while (avoidSet.contains(candidateID)) {
+      builder.append('.')
+      candidateID = builder.toString()
+    }
+    ModuleID(candidateID)
   }
 
   /** Creates a digest-based generator whose prefix is not a prefix of any of
    *  the IDs in [[avoid]] nor of the public module IDs.
    */
   def makeFromDigestGenerator(avoid: Iterable[ModuleID]): FromDigestGenerator = {
-    val freeInternalPrefix = Iterator
-      .iterate("internal-")(_ + "-")
-      .find(p => !avoid.exists(_.id.startsWith(p)) && !publicModuleIDs.exists(_.id.startsWith(p)))
-      .get
-    new FromDigestGenerator(freeInternalPrefix)
+    /* Here we can use `equalsIgnoreCase`, even though it has a poor notion of
+     * case folding (which is even Char-based, not code point-based). That is
+     * because we always compare against a string of the form 'internal---' for
+     * an arbitrary number of '-'.
+     *
+     * - Only '-' is equal to '-'
+     * - Only 'i ı I İ' are equal to 'i'
+     * - Only ASCII letters are equal to the other letters of "internal"
+     *
+     * All these cases are handled by `equalsIgnoreCase`.
+     */
+    val DashIndex = 8
+    var prefix = "internal-"
+    for (moduleID <- avoid.iterator ++ publicModuleIDs.iterator) {
+      val id = moduleID.id
+      while (id.length() >= prefix.length() &&
+          id.charAt(DashIndex) == '-' && // fast exit
+          id.substring(0, prefix.length()).equalsIgnoreCase(prefix)) {
```

Comment:
I feel this is a bit an odd mix between fast and short: If we've identified a conflict, we'll re check the "fast exit" every time and we'll also re-compare the beginning of the prefix every time (even though we just appended a dash).

Have you considered:

```scala
for (moduleID <- avoid.iterator ++ publicModuleIDs.iterator) {
  val id = moduleID.id
  val idLen = id.length()
  val pLen = prefix.length()
  if (idLen >= pLen &&
      id.charAt(DashIndex) == '-' && // fast exit
      id.substring(0, pLen).equalsIgnoreCase(prefix)) {
    // append dashes until there is no prefix conflict anymore.
    val lastDash = id.lastIndexOf('-', pLen)
    prefix += "-" * (lastDash + 1 - pLen) // TODO: check for off-by-one
  }
}

```